### PR TITLE
trivial: uefi-self-test: Skip TPM2 PCR0 test if missing permissions

### PR DESCRIPTION
**Pull request by superm1**
A simple test for the existence of the node `/sys/class/tpm/tpm0` is
not enough to indicate that fwupd will be able to use the TPM.

Check the permissions of the devnode to make sure the underlying
library can access it.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/uefi/fu-self-test.c [hunk 1]
```diff
@@ -44,18 +46,18 @@ static void
...
 	g_autoptr(GPtrArray) pcrXs = NULL;
+	gboolean physical_tpm;
 
-	sysfstpmdir = fu_common_get_path (FU_PATH_KIND_SYSFSDIR_TPM);
-	devpath = g_build_filename (sysfstpmdir, "tpm0", NULL);
-
-	if (!g_file_test (devpath, G_FILE_TEST_EXISTS) && (tpm_server_running == NULL)) {
+	devpath = g_build_filename ("/dev", "tpm0", NULL);
+	physical_tpm = g_file_test (devpath, G_FILE_TEST_EXISTS) &&
+		       g_access (devpath, W_OK | R_OK) == 0;
```
Comment (hughsie) [inline plugins/uefi/fu-self-test.c]: that works, but it's a bit of a cludge. can't we do a `g_error_matches()` on the error instead?

Comment (hughsie) [reply plugins/uefi/fu-self-test.c]: Can't we map the output of `Esys_Initialize()` to differing GError codes? At the moment they're all mapped to `G_IO_ERROR_NOT_SUPPORTED`

Comment (diabonas) [reply plugins/uefi/fu-self-test.c]: When the second argument of `Esys_Initialize` is `NULL`, [different libraries](https://github.com/tpm2-software/tpm2-tss/blob/master/src/tss2-tcti/tctildr-dl.c#L33-L54) are tried until one that works is found. Its return value is `TSS2_ESYS_RC_SUCCESS` if everything worked as expected, or `TSS2_TCTI_RC_IO_ERROR` if initialisation failed, no further information on the nature of the error (missing libraries/permission problems/TPM error/...) is available. It is possible to use the second argument to specify through which library the TPM should be accessed, but usually it is best to leave this to tpm2-tss.

Checking whether we have permission to communicate with the TPM is not easy: first of all, both `/dev/tpm0` and `/dev/tpmrm0` can be used to access it, so we would need to check if we have RW permissions for either of these. But even if the current user doesn't have permissions to access these devices, tpm2-abrmd might be installed and running as the `tss` user, which means that we don't have direct access to the TPM at all, but can still talk to it using the resource manager.

To test whether it makes sense to run the test, we could try running `Esys_Initialize` and see whether that succeeds:
```
TSS2_RC rc;
g_autoptr(ESYS_CONTEXT) ctx = NULL;
rc = Esys_Initialize (&ctx, NULL, NULL);
if (rc != TSS2_RC_SUCCESS) {
	g_test_skip ("no physical or simulated TPM 2.0 device available");
	return;
}
```
If that fails, there is definitely something wrong on the tpm2-tss side, so it does not make sense to run the test.

Comment (hughsie) [reply plugins/uefi/fu-self-test.c]: I like this version.

**Comment by diabonas** [comment]
`export TPM_SERVER_RUNNING=1` in [`contrib/ci/arch.sh`](https://github.com/fwupd/fwupd/blob/d1681e6de558eca6aad42adcc2c1c47534972d4a/contrib/ci/arch.sh#L23) can now be removed as well.

**Comment by superm1** [comment]
> export TPM_SERVER_RUNNING=1 in contrib/ci/arch.sh can now be removed as well.

You know actually I think something else needs to be put in place to check that variable again in the test case.  It's a special case when it's running that you want to make sure that it actually passes in CI.

**Comment by diabonas** [comment]
> You know actually I think something else needs to be put in place to check that variable again in the test case. It's a special case when it's running that you want to make sure that it actually passes in CI.

Makes sense :+1:

**Pull request merged**